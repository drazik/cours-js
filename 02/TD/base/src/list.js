/*
 * @typedef {Object} Item
 * @property {function} addItem - Permet d'ajouter un item à la liste
 */

/**
 * Initialise la liste
 *
 * @param {HTMLUListElement} listElement - La liste d'items
 *
 * @return {Item}
 */
export const initList = (listElement) => {
  /**
   * Crée un nouvel item et l'ajoute à la liste
   *
   * @param {string} label
   */
  const addItem = (label) => {
    // 👉 Utiliser la fonction `makeItem` pour créer un élément représentant
    // un item de la liste à partir du label

    // 👉 Ajouter l'item à la liste
  }

  return { addItem }
}

/**
 * Initialise un item de la liste
 *
 * @return {HTMLLIElement} L'élément du DOM de l'item
 */
const makeItem = (label) => {
  // 👉 Créer l'élément de l'item à l'aide de la fonction `createItemNode`
  // const element = ...

  // 👉 Récupérer l'input et le bouton dans l'élément créé
  // const input = ...
  // const button = ...

  const init = () => {
    input.addEventListener("change", handleSelect)
    button.addEventListener("click", destroy)
  }

  /**
   * Gère l'état visuel "sélectionné" ou non de l'item.
   */
  const handleSelect = (e) => {
    // 👉 Ajouter la classe list__item--selected à l'item si e.target.checked
    // vaut true. La retirer sinon
  }

  const destroy = () => {
    // 👉 Supprimer l'élément du DOM

    input.removeEventListener("change", handleSelect)
    button.addEventListener("click", destroy)
  }

  init()

  return element
}

/**
 * Crée un nouvel élément DOM pour un item de la liste
 *
 * @param {string} itemLabel - Le label de l'item
 *
 * @return {HTMLLiElement} L'élément créé
 */
const createItemElement = (itemLabel) => {
  // 👉 Créer un élément de type "li"

  // 👉 Affecter la classe "list__item" à l'élément

  // 👉 Créer un élément de type "label"

  // 👉 Créer un élément de type "input"

  // 👉 Affecter le type "checkbox" à l'input

  // 👉 Créer un élément de type "button"

  // 👉 Affecter le type "button" au button

  // 👉 Affecter la classe "list__remove-btn" au button

  // 👉 Ajouter la chaîne "❌" dans le button

  // 👉 Affecter la valeur 'Supprimer' à l'attribut aria-label du button

  // 👉 Ajouter l'input dans le label

  // 👉 Ajouter itemLabel dans le label

  // 👉 Ajouter le label dans l'item

  // 👉 Ajouter le button dans l'item
}
