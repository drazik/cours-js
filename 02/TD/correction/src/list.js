/*
 * @typedef {Object} Item
 * @property {function} addItem - Permet d'ajouter un item à la liste
 */

/**
 * Initialise la liste
 *
 * @param {HTMLUListElement} listElement - La liste d'items
 *
 * @return {Item}
 */
export const initList = (listElement) => {
  /**
   * Crée un nouvel item et l'ajoute à la liste
   *
   * @param {string} label
   */
  const addItem = (label) => {
    // 👉 Utiliser la fonction `makeItem` pour créer un élément représentant
    // un item de la liste à partir du label
    const item = makeItem(label)

    // 👉 Ajouter l'item à la liste
    listElement.append(item)
  }

  return { addItem }
}

/**
 * Initialise un item de la liste
 *
 * @return {HTMLLIElement} L'élément du DOM de l'item
 */
const makeItem = (label) => {
  // 👉 Créer l'élément de l'item à l'aide de la fonction `createItemNode`
  const element = createItemElement(label)

  // 👉 Récupérer l'input et le bouton dans l'élément créé
  const input = element.querySelector("input")
  const button = element.querySelector("button")

  const init = () => {
    input.addEventListener("change", handleSelect)
    button.addEventListener("click", destroy)
  }

  /**
   * Gère l'état visuel "sélectionné" ou non de l'item.
   */
  const handleSelect = (e) => {
    // 👉 Ajouter la classe list__item--selected à l'item si e.target.checked
    // vaut true. La retirer sinon
    const selectedClass = "list__item--selected"

    if (e.target.checked) {
      element.classList.add(selectedClass)
    } else {
      element.classList.remove(selectedClass)
    }
  }

  const destroy = () => {
    // 👉 Supprimer l'élément du DOM
    element.remove()

    input.removeEventListener("change", handleSelect)
    button.addEventListener("click", destroy)
  }

  init()

  return element
}

/**
 * Crée un nouvel élément DOM pour un item de la liste
 *
 * @param {string} itemLabel - Le label de l'item
 *
 * @return {HTMLLiElement} L'élément créé
 */
const createItemElement = (itemLabel) => {
  // 👉 Créer un élément de type "li"
  const item = document.createElement("li")

  // 👉 Affecter la classe "list__item" à l'élément
  item.classList.add("list__item")

  // 👉 Créer un élément de type "label"
  const label = document.createElement("label")

  // 👉 Créer un élément de type "input"
  const input = document.createElement("input")

  // 👉 Affecter le type "checkbox" à l'input
  input.type = "checkbox"

  // 👉 Créer un élément de type "button"
  const button = document.createElement("button")

  // 👉 Affecter le type "button" au button
  button.type = "button"

  // 👉 Affecter la classe "list__remove-btn" au button
  button.className = "list__remove-btn"

  // 👉 Ajouter la chaîne "❌" dans le button
  button.textContent = "❌"

  // 👉 Affecter la valeur 'Supprimer' à l'attribut aria-label du button
  button.setAttribute("aria-label", "Supprimer")

  // 👉 Ajouter l'input dans le label
  label.append(input)

  // 👉 Ajouter itemLabel dans le label
  label.append(itemLabel)

  // 👉 Ajouter le label dans l'item
  item.append(label)

  // 👉 Ajouter le button dans l'item
  item.append(button)

  return item
}
