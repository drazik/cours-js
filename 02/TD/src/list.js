/**
 * Crée un nouvel élément DOM pour un item de la liste
 *
 * @param {string} itemLabel - Le label de l'item
 *
 * @return {HTMLLiElement} L'élément créé
 */
export function createItemNode(itemLabel) {
  // 👉 Créer un élément de type li
  // let item = ...

  // 👉 Affecter la classe list__item à l'élément
  // item.className = ...

  // 👉 Créer un élément de type label
  // let label = ...
  
  // 👉 Créer un élément de type input
  // let input = ...

  // 👉 Affecter le type checkbox à l'input
  // input.type = ...

  // 👉 Créer un élément de type button
  // let button = ...

  // 👉 Affecter le type button au button

  // 👉 Affecter la classe list__remote-btn au button

  // 👉 Ajouter la chaîne '❌' dans le button
  // button.append(...)

  // 👉 Affecter la valeur 'Supprimer' à l'attribut aria-label du button

  // 👉 Ajouter l'input dans le label
  // label.append(...)

  // 👉 Ajouter itemLabel dans le label

  // 👉 Ajouter le label dans l'item

  // 👉 Ajouter le button dans l'item

  /**
   * Ajoute ou supprime l'état sélectionné à l'item
   *
   * @param {Event} e
   */
  function handleSelect(e) {
    // 👉 Ajouter la classe list__item--selected à l'item si e.target.value
    // vaut true. La retirer sinon
  }

  /**
   * Supprime l'item
   */
  function handleRemove() {
    // 👉 Supprimer l'item du DOM
  }

  input.addEventListener('change', handleSelect)
  button.addEventListener('click', handleRemove)


  return item
}

/**
 * Initialise la liste
 *
 * @param {HTMLUListElement} listElement - La liste d'items
 *
 * @return {Object}
 */
export function initList(listElement) {
  /**
   * Crée un nouvel item et l'ajoute à la liste
   *
   * @param {string} label
   */
  function addItem(label) {
    // 👉 Créer l'item en appelant la fonction createItemNode en lui passant le
    // label
    // let item = ...

    // 👉 Ajouter l'item à la liste
  }

  return { addItem }
}
