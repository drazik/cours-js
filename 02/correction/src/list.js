/**
 * Crée un nouvel élément DOM pour un item de la liste
 *
 * @param {string} itemLabel - Le label de l'item
 *
 * @return {HTMLLiElement} L'élément créé
 */
export function createItemNode(itemLabel) {
  // 👉 Créer un élément de type li
  let item = document.createElement('li')

  // 👉 Affecter la classe list__item à l'élément
  item.className = 'list__item'

  // 👉 Créer un élément de type label
  let label = document.createElement('label')
  
  // 👉 Créer un élément de type input
  let input = document.createElement('input')

  // 👉 Affecter le type checkbox à l'input
  input.type = 'checkbox'

  // 👉 Créer un élément de type button
  let button = document.createElement('button')

  // 👉 Affecter le type button au button
  button.type = 'button'

  // 👉 Affecter la classe list__remote-btn au button
  button.className = 'list__remove-btn'

  // 👉 Ajouter la chaîne '❌' dans le button
  button.append('❌')

  // 👉 Affecter la valeur 'Supprimer' à l'attribut aria-label du button
  button.setAttribute('aria-label', 'Supprimer')

  // 👉 Ajouter l'input dans le label
  label.append(input)

  // 👉 Ajouter itemLabel dans le label
  label.append(itemLabel)

  // 👉 Ajouter le label dans l'item
  item.append(label)

  // 👉 Ajouter le button dans l'item
  item.append(button)

  /**
   * Ajoute ou supprime l'état sélectionné à l'item
   *
   * @param {Event} e
   */
  function handleSelect(e) {
    // 👉 Ajouter la classe list__item--selected à l'item si e.target.value
    // vaut true. La retirer sinon
    item.classList.toggle('list__item--selected', e.target.checked)
  }

  /**
   * Supprime l'item
   */
  function handleRemove() {
    // 👉 Supprimer l'item du DOM
    item.remove()
  }

  input.addEventListener('change', handleSelect)
  button.addEventListener('click', handleRemove)


  return item
}

/**
 * Initialise la liste
 *
 * @param {HTMLUListElement} listElement - La liste d'items
 *
 * @return {Object}
 */
export function initList(listElement) {
  /**
   * Crée un nouvel item et l'ajoute à la liste
   *
   * @param {string} label
   */
  function addItem(label) {
    // 👉 Créer l'item en appelant la fonction createItemNode en lui passant le
    // label
    let item = createItemNode(label)

    // 👉 Ajouter l'item à la liste
    listElement.append(item)
  }

  return { addItem }
}
