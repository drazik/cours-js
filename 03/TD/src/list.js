/**
 * Cr√©e un nouvel √©l√©ment DOM pour un item de la liste
 *
 * @param {string} itemLabel - Le label de l'item
 *
 * @return {HTMLLiElement} L'√©l√©ment cr√©√©
 */
export function createItemNode(itemLabel) {
  let item = document.createElement('li')
  item.className = 'list__item js-list-item'

  let label = document.createElement('label')
  
  let input = document.createElement('input')
  input.type = 'checkbox'
  input.className = 'js-list-item-checkbox'

  let button = document.createElement('button')
  button.type = 'button'
  button.className = 'list__remove-btn js-list-item-remove'
  button.append('‚ùå')
  button.setAttribute('aria-label', 'Supprimer')

  label.append(input)
  label.append(itemLabel)

  item.append(label)
  item.append(button)

  /**
   * Ajoute ou supprime l'√©tat s√©lectionn√© √† l'item
   *
   * @param {Event} e
   */
  function handleSelect(e) {
    item.classList.toggle('list__item--selected', e.target.checked)
  }

  /**
   * Supprime l'item
   */
  function handleRemove() {
    item.remove()
  }

  // üëâ Supprimer ces event listeners
  input.addEventListener('change', handleSelect)
  button.addEventListener('click', handleRemove)


  return item
}

/**
 * Initialise la liste
 *
 * @param {HTMLUListElement} listElement - La liste d'items
 *
 * @return {Object}
 */
export function initList(listElement) {
  /**
   * Cr√©e un nouvel item et l'ajoute √† la liste
   *
   * @param {string} label
   */
  function addItem(label) {
    let item = createItemNode(label)

    listElement.append(item)

    return item
  }

  // üëâ Ajouter une d√©l√©gation d'√©v√©nement sur `listElement`, pour l'√©v√©nement
  // `change`, sur les √©l√©ments qui ont la classe `js-list-item-checkbox`
  // ...


  // üëâ Ajouter une d√©l√©gation d'√©v√©nement sur `listElement`, pour l'√©v√©nement
  // `click`, sur les √©l√©ments qui ont la classe `js-list-item-remove`
  // ...

  return { addItem }
}
