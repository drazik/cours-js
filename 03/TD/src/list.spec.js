import { createItemNode, initList } from './list'
import { getByLabelText, fireEvent } from '@testing-library/dom'
import '@testing-library/jest-dom/extend-expect'

describe('createItemNode', () => {
  it('should create an item with the given label', () => {
    let item = createItemNode('huile')

    expect(item).toMatchSnapshot()
  })
})

describe('initList', () => {
  it('should return an object exposing an `addItem` function that', () => {
    let root = document.createElement('ul')
    let list = initList(root)

    expect(list.addItem).toBeInstanceOf(Function)

    list.addItem('farine')

    expect(getByLabelText(root, 'farine')).toBeTruthy()
  })

  it('should handle item removal', () => {
    let root = document.createElement('ul')
    let list = initList(root)

    list.addItem('farine')

    let removeBtn = getByLabelText(root, /supprimer/i)

    fireEvent.click(removeBtn)

    expect(() => getByLabelText(root, 'farine')).toThrow()
  })

  it('should handle item selection', () => {
    let root = document.createElement('ul')
    let list = initList(root)
    let item = list.addItem('farine')

    let checkbox = getByLabelText(item, 'farine')
    fireEvent.click(checkbox)

    expect(item).toHaveClass('list__item--selected')

    fireEvent.click(checkbox)
    expect(item).not.toHaveClass('list__item--selected')
  })
})
