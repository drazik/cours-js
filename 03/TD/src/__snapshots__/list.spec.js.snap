// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createItemNode should create an item with the given label 1`] = `
<li
  class="list__item js-list-item"
>
  <label>
    <input
      class="js-list-item-checkbox"
      type="checkbox"
    />
    huile
  </label>
  <button
    aria-label="Supprimer"
    class="list__remove-btn js-list-item-remove"
    type="button"
  >
    ‚ùå
  </button>
</li>
`;
